Add retries to all resources and datasources

mitchnielsen opened the pull request:
Uses the [hashicorp/retryablehttp](https://pkg.go.dev/github.com/hashicorp/go-retryablehttp#section-readme) package to get an HTTP client with
retries and exponential backoff out of the box. This is implemented
within our client package, meaning calls to the CRUD methods from both
resources and datasources will inherit that functionality.

At a high level, the changes here are:
- Replacing the `http` client with a `retryablehttp` client
- Reverting implementations of `avast/retry` in resources
- Removing `avast/retry`-related helpers

Related to https://linear.app/prefect/issue/PLA-183/add-retries-to-resources

mitchnielsen committed:
Work Pool resource: retry Create and Read

Adds retries for Create and Read on the Work Pool resource.

mitchnielsen commented:
It probably makes sense just to implement retries on all resources/datasources at once, rather than incrementally adding them where we get reports of failures - I'm going to see if I can get Cursor to do most of the heavy lifting there, and if we can add a facility to make it easier to add/tweak in the future 🤔

parkedwards commented:
@mitchnielsen trying to think, are there calls that we WOULDNT want to retry on?

mitchnielsen commented:
> trying to think, are there calls that we WOULDNT want to retry on?

Asking in Slack 👌🏼

mitchnielsen committed:
Use hashicorp/retryablehttp for retries

Uses the hashicorp/retryablehttp package to get an HTTP client with
retries and exponential backoff out of the box. This is implemented
within our client package, meaning calls to the CRUD methods from both
resources and datasources will inherit that functionality.

mitchnielsen committed:
Remove unused client functions

mitchnielsen committed:
Expand comment for retryablehttp usage

mitchnielsen committed:
Exit on error for blockSchemaClient.List

mitchnielsen committed:
Merge branch 'main' of github.com:PrefectHQ/terraform-provider-prefect into add-more-retries

mitchnielsen committed:
Increase the retry interval and maximum

Blocks, and other resources, need more time to resolve objects so let's
bump up the defaults to accommodate that.

mitchnielsen committed:
Bump min retry wait seconds to 3

Still saw failures in acceptance tests in CI

mitchnielsen committed:
Bump retry seconds to 3

Was still seeing errors in the acceptance tests

mitchnielsen committed:
Remove time/retry overrides, add retry check func

- Adds a function to define how to determine whether or not to retry,
  extending the default function so we can accept 409s and retry 404s.
  We can adapt this function over time as well.
- Removes the time and retry overrides, as those were not the reason
  that block tests were failing - the requests themselves were passing,
  and didn't need retries - what they needed was to retry 404s and not
  retry 409s.

mitchnielsen committed:
Use the stdlib http.Client

Uses the stdlib http.Client interface so we don't need to specify the
retryablehttp variant in every client method.

mitchnielsen commented on internal/client/client.go:
```diff
@@ -10,14 +11,38 @@ import (
+
 	client := &Client{
-		hc: http.DefaultClient,
+		hc: httpClient,
```
This is the main change here - we're replacing the default HTTP client with our retry-able HTTP client.

mitchnielsen commented on internal/client/client.go:
```diff
@@ -36,26 +61,6 @@ func New(opts ...Option) (*Client, error) {
 	return client, nil
 }
 
-// MustNew returns a new client or panics if an error occurred.
-func MustNew(opts ...Option) *Client {
-	client, err := New(opts...)
-	if err != nil {
-		panic(fmt.Sprintf("error occurred during construction: %s", err))
-	}
-
-	return client
-}
-
-// WithClient configures the underlying http.Client used to send
-// requests.
-func WithClient(httpClient *http.Client) Option {
-	return func(client *Client) error {
-		client.hc = httpClient
-
-		return nil
-	}
-}
-
```
These were unused.

mitchnielsen commented on internal/client/client.go:
```diff
@@ -97,3 +102,21 @@ func WithDefaults(accountID uuid.UUID, workspaceID uuid.UUID) Option {
 		return nil
 	}
 }
+
+func checkRetryPolicy(ctx context.Context, resp *http.Response, err error) (bool, error) {
+	// If the response is a 409 (StatusConflict), that means the request
+	// eventually succeeded and we don't need to make the request again.
+	if resp.StatusCode == http.StatusConflict {
+		return false, nil
+	}
+
+	// If the response is a 404 (NotFound), try again. This is particularly
+	// relevant for block-related objects that are created asynchronously.
+	if resp.StatusCode == http.StatusNotFound {
+		return true, err
+	}
+
+	// Fall back to the default retry policy for any other status codes.
+	//nolint:wrapcheck // we've extended this method, no need to wrap error
+	return retryablehttp.DefaultRetryPolicy(ctx, resp, err)
+}
```
Here we can be specific about what warrants a retry. This is cool because we knew in the past we didn't need to retry 409s, but the retry implementation we had at the time didn't really let us achieve that.

mitchnielsen commented on internal/client/client.go:
```diff
@@ -10,14 +11,38 @@ import (
+	// Finally, convert the retryablehttp client to a standard http client.
+	// This allows us to retain the `http.Client` interface, and avoid specifying
+	// the `retryablehttp.Client` interface in our client methods.
+	httpClient := retryableClient.StandardClient()
```
This part is pretty sweet. It saves a lot of lines changed - see https://github.com/PrefectHQ/terraform-provider-prefect/pull/286/commits/f86254b42555e37a760580ecda01a4705b52d009.

parkedwards reviewed (approved):
awesome!
